### PR TITLE
Fix multiple spans being created for HTTPUrlConnection HEAD requests

### PR DESCRIPTION
**Pull request by JonasKunz**
## What does this PR do?

While testing with a local spring-boot app I noticed that out instrumentation created 14 spans per single RestTemplate request.

This issue occured because I instructed the `RestTemplate` to ignore the response body by setting the expected type to `Void.class`. This causes spring to send `HEAD` requests instead. And as it turns out, our `HTTPUrlConnection` instrumentation doesn't deal well with those:

The internal `connected` field is not switched to `true` for `HEAD` requests, causing multiple spans to be created when invoking `getResponseCode` multiple times.

This PR fixes this by also checking the `responseCode` field to infer whether the `HTTPUrlConnection` has already connected or not.

## Checklist

- [x] This is a bugfix
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [x] I have added tests that would fail without this fix

### CODE REVIEW

File: apm-agent-plugins/apm-urlconnection-plugin/src/main/java/co/elastic/apm/agent/urlconnection/HttpUrlConnectionInstrumentation.java [hunk 1]
```diff
@@ -79,15 +79,19 @@ public static class AdviceClass {
             @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
             public static Object enter(@Advice.This HttpURLConnection thiz,
                                        @Advice.FieldValue("connected") boolean connected,
+                                       @Advice.FieldValue("responseCode") int responseCode,
                                        @Advice.Origin String signature) {
 
+                //With HEAD requests the connected stays false
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-urlconnection-plugin/src/main/java/co/elastic/apm/agent/urlconnection/HttpUrlConnectionInstrumentation.java:85]: [minor] that's really an implementation detail here, could we also check the HTTP verb to be sure ?

**Review by SylvainJuge — APPROVED** [review]
Good catch !

File: apm-agent-plugins/apm-urlconnection-plugin/src/main/java/co/elastic/apm/agent/urlconnection/HttpUrlConnectionInstrumentation.java (hunk 1 above)
Comment (JonasKunz) [reply apm-agent-plugins/apm-urlconnection-plugin/src/main/java/co/elastic/apm/agent/urlconnection/HttpUrlConnectionInstrumentation.java:85]: I'd argue that even for non-HEAD requests this approach is safer: If a responseCode is available, this implies that the connection has already been made. Therefore I think an additional check would not be good here.

Comment (SylvainJuge) [reply apm-agent-plugins/apm-urlconnection-plugin/src/main/java/co/elastic/apm/agent/urlconnection/HttpUrlConnectionInstrumentation.java:85]: Good for me, let's keep it as-is then !

**Comment by SylvainJuge** [comment]
@Mergifyio refresh

**Pull request merged**